ql/src/java/org/apache/hadoop/hive/ql/history/HiveScribeImpl.java

```diff
@@ -260,13 +261,16 @@ public void snapshotTaskProgress(RecordTypes recordTypes, QueryStats stats, Map<
     insertTaskProgress(stats.getTaskProgress(), timeStamp, snapshot.toString());
   }
 
-  private void insertTaskProgress(Map<String, String> taskProgressStats, String timeStamp, String taskProgress) {
-    for (Map.Entry<String, String> ent : taskProgressStats.entrySet()) {
-      if (taskProgress.equals(ent.getValue())) {
-        return;
-      }
+  private void insertTaskProgress(ArrayList<QueryStats.progressSnapshot> taskProgressStats, Long timeStamp, String taskProgress) {
+    int listSize = taskProgressStats.size();
+    if ( listSize == 0 || taskProgress.equals(taskProgressStats.get(listSize - 1).getValue())) {
```

Comment:
BUG.

If the list is empty, why are you skipping this event??? This will end up skipping every single `progress` event and you get an empty array.

Please test and verify it carefully.